clients/src/main/java/org/apache/kafka/clients/consumer/internals/MemberState.java

```diff
@@ -17,57 +17,113 @@
+     * received from the broker)
+     */
+    ACKNOWLEDGING,
+
+    /**
+     * Member is active in a group and has processed all assignments received. While in this
+     * state, the member will send heartbeats on the interval.
      */
     STABLE,
 
     /**
-     * Member transitions to this state when it receives a
-     * {@link org.apache.kafka.common.protocol.Errors#UNKNOWN_MEMBER_ID} or
-     * {@link org.apache.kafka.common.protocol.Errors#FENCED_MEMBER_EPOCH} error from the
-     * broker. This is a recoverable state, where the member
-     * gives up its partitions by invoking the user callbacks for onPartitionsLost, and then
-     * transitions to {@link #UNJOINED} to rejoin the group as a new member.
+     * Member transitions to this state when it receives a {@link Errors#UNKNOWN_MEMBER_ID} or
+     * {@link Errors#FENCED_MEMBER_EPOCH} error from the broker, indicating that it has been
+     * left out of the group. While in this state, the member will stop sending heartbeats, it
+     * will give up its partitions by invoking the user callbacks for onPartitionsLost, and then
+     * transition to {@link #JOINING} to re-join the group as a new member.
      */
     FENCED,
 
     /**
-     * The member failed with an unrecoverable error
+     * The member transitions to this state after a call to unsubscribe. While in this state, the
+     * member will stop sending heartbeats, will commit offsets if needed and release its
+     * assignment (calling user's callback for partitions revoked or lost). When all these
+     * actions complete, the member will transition out of this state into {@link #LEAVING} to
+     * effectively leave the group.
      */
-    FAILED;
+    PREPARE_LEAVING,
 
+    /**
+     * Member has committed offsets and releases its assignment, so it stays in this state until
+     * the next heartbeat request is sent out with epoch -1 to effectively leave the group. This
+     * state indicates that the next heartbeat request must be sent without waiting for the
+     * heartbeat interval to expire.
+     */
+    LEAVING,
+
+    /**
+     * The member failed with an unrecoverable error received in a heartbeat response. This in an
+     * unrecoverable state where the member won't send any requests to the broker and cannot
+     * perform any other transition.
+     */
+    FATAL;
+
+    /**
+     * Valid state transitions
+     */
     static {
-        // Valid state transitions
-        STABLE.previousValidStates = Arrays.asList(UNJOINED, RECONCILING);
 
-        RECONCILING.previousValidStates = Arrays.asList(STABLE, UNJOINED);
+        STABLE.previousValidStates = Arrays.asList(JOINING, ACKNOWLEDGING);
+
+        RECONCILING.previousValidStates = Arrays.asList(STABLE, JOINING, ACKNOWLEDGING);
+
+        ACKNOWLEDGING.previousValidStates = Arrays.asList(RECONCILING);
+
+        FATAL.previousValidStates = Arrays.asList(JOINING, STABLE, RECONCILING, ACKNOWLEDGING);
```

Comment:
There is no transition from prepare leaving to fatal with the current usage of fatal (only when receiving fatal errors in HB response), because we don't send HB while in prepare leaving. 

As for leaving, I would say we shouldn't either, because even if it is a state where we do send HB, we transition out of it as soon as the HB request is ready to be sent (without waiting for the actual send or any response). That being said, this makes me realize that the same reasoning applies for ACKNOWLEDGING, there shouldn't be a way of transitioning from ACK to FATAL, just because we transition out of it on heartbeat sent.

Reply:
> There is no transition from prepare leaving to fatal with the current usage of fatal (only when receiving fatal errors in HB response), because we don't send HB while in prepare leaving.

We don't send an explicit HB while in prepare leaving. However, we continue to heartbeat so I assume that we could receive an error or be fenced while in this state. Regarding the ACK case, I think that we could receive a similar response while in ACK so the same applies. Do you agree?